modules/mapper-extras/src/main/java/org/elasticsearch/index/query/ScoreScriptUtils.java

```diff
@@ -0,0 +1,218 @@
+            BytesRef value = dvs.getEncodedValue();
+            if (value == null) return 0;
+            int[] docDims = VectorEncoderDecoder.decodeSparseVectorDims(value);
+            float[] docValues = VectorEncoderDecoder.decodeSparseVector(value);
+            return intDotProductSparse(queryValues, queryDims, docValues, docDims);
+        }
+    }
+
+    /**
+     * Calculate cosine similarity between a query's sparse vector and documents' sparse vectors
+     *
+     * CosineSimilaritySparse is implemented as a class to use
+     * painless script caching to prepare queryVector and calculate queryVectorMagnitude
+     * only once per script execution for all documents.
+     * A user will call `cosineSimilaritySparse(params.queryVector, doc['my_vector'])`
+     */
+    public static final class CosineSimilaritySparse {
+        double[] queryValues;
+        int[] queryDims;
+        double queryVectorMagnitude;
+
+        // prepare queryVector once per script execution
+        public CosineSimilaritySparse(Map<String, Number> queryVector) {
+            //break vector into two arrays dims and values
+            int n = queryVector.size();
+            queryValues = new double[n];
+            queryDims = new int[n];
+            double dotProduct = 0;
+            int i = 0;
+            for (Map.Entry<String, Number> dimValue : queryVector.entrySet()) {
+                try {
+                    queryDims[i] = Integer.parseInt(dimValue.getKey());
+                } catch (final NumberFormatException e) {
+                    throw new IllegalArgumentException("Failed to parse a query vector dimension, it must be an integer!", e);
+                }
+                queryValues[i] = dimValue.getValue().doubleValue();
+                dotProduct +=  queryValues[i] *  queryValues[i];
+                i++;
+            }
+            this.queryVectorMagnitude = Math.sqrt(dotProduct);
+            // Sort dimensions in the ascending order and sort values in the same order as their corresponding dimensions
+            sortSparseDimsDoubleValues(queryDims, queryValues, n);
+        }
+
+        public double cosineSimilaritySparse(VectorScriptDocValues.SparseVectorScriptDocValues dvs) {
+            BytesRef value = dvs.getEncodedValue();
+            if (value == null) return 0;
+            int[] docDims = VectorEncoderDecoder.decodeSparseVectorDims(value);
+            float[] docValues = VectorEncoderDecoder.decodeSparseVector(value);
+
+            // calculate docVector magnitude
+            double dotProduct = 0;
+            for (float docValue : docValues) {
+                dotProduct += (double) docValue * docValue;
+            }
+            final double docVectorMagnitude = Math.sqrt(dotProduct);
+
+            double docQueryDotProduct = intDotProductSparse(queryValues, queryDims, docValues, docDims);
+            return docQueryDotProduct / (docVectorMagnitude * queryVectorMagnitude);
+        }
+    }
```

Comment:
can you add unit tests to all public methods in this class?

Reply:
@jpountz  Thanks Adrien, I have added the tests; the PR is ready for another round of review.